### DISCUSSION TITLE
Composition order, defaults list

### DISCUSSION THREAD

**Discussion by DirkKuhn** [mit-ll-responsible-ai/hydra-zen#473] [General]
Hi,

I noticed the following gotcha:
```python
from typing import Optional
from hydra_zen import store, zen

class Test:
    def __init__(self, number: int):
        self.number = number

    def __repr__(self) -> str:
        return f'Test(number={self.number})'

def app(test: Optional[Test] = None):
    print(f'{test=}')

test_store = store(group='test')
test_store(Test, number=0)

store(app, hydra_defaults=[{'test': 'Test'}, '_self_'], name='config')

store.add_to_hydra_store()

if __name__ == '__main__':
    zen(app).hydra_main(config_name='config', config_path=None, version_base='1.3')
```
prints
```
test=None
```
By putting `'_self_'` last in the defaults list the default value `test=None` from `app`'s signature is added at the end, setting `test` to `None`. This centainly did confuse me.
Should the values from the signature and the values from the config be handled differently or is this behaviour actually wanted?

If `app`'s config is changed to
```python
store(app, hydra_defaults=[{'test': 'Test'}, '_self_'], test=dict(number=10), name='config')
```
everything works as I would expect it. The output is
```
test=Test(number=10)
```

**Comment by rsokl** [comment]
That is the expected behavior when you include `_self_` as the last element in `hydra_defaults`. You will typically want `_self_` to be the first element. See: https://hydra.cc/docs/tutorials/structured_config/defaults/

If you do `store(app, hydra_defaults=['_self_', {'test': 'Test'}], name='config')`, running your original program prints `test=Test(number=0)`
